src/Bridge/Doctrine/Orm/SubcollectionDataProvider.php

```diff
@@ -0,0 +1,128 @@
+        $this->collectionExtensions = $collectionExtensions;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws RuntimeException
+     */
+    public function getSubcollection(string $resourceClass, array $identifiers, array $context, string $operationName)
+    {
+        $manager = $this->managerRegistry->getManagerForClass($resourceClass);
+        if (null === $manager) {
+            throw new ResourceClassNotSupportedException();
+        }
+
+        $repository = $manager->getRepository($resourceClass);
+        if (!method_exists($repository, 'createQueryBuilder')) {
+            throw new RuntimeException('The repository class must have a "createQueryBuilder" method.');
+        }
+
+        $queryBuilder = $repository->createQueryBuilder('o');
+        $queryNameGenerator = new QueryNameGenerator();
+        $previousQueryBuilder = null;
+        $previousAlias = null;
+
+        $num = count($context['identifiers']);
+
+        while ($num--) {
+            list($identifier, $identifierResourceClass) = $context['identifiers'][$num];
+            $previousIdentifier = $context['identifiers'][$num + 1][0] ?? $context['property'];
+
+            $classMetadata = $manager->getClassMetadata($identifierResourceClass);
+
+            $qb = $manager->createQueryBuilder();
+            $alias = $queryNameGenerator->generateJoinAlias($identifier);
+
+            if (null !== $previousIdentifier) {
+                //MANY_TO_MANY relations needs an explicit join so that the identifier part can be retrieved
+                if ($classMetadata->getAssociationMapping($previousIdentifier)['type'] === ClassMetadataInfo::MANY_TO_MANY) {
+                    $joinAlias = $queryNameGenerator->generateJoinAlias($previousIdentifier);
+
+                    $qb->select($joinAlias)
+                        ->from($identifierResourceClass, $alias)
+                        ->innerJoin("$alias.$previousIdentifier", $joinAlias);
+                } else {
+                    $qb->select("IDENTITY($alias.$previousIdentifier)")
+                        ->from($identifierResourceClass, $alias);
+                }
+            }
+
+            $normalizedIdentifiers = $this->normalizeIdentifiers($identifiers[$identifier], $manager, $identifierResourceClass);
+
+            foreach ($normalizedIdentifiers as $key => $value) {
+                $placeholder = $queryNameGenerator->generateParameterName($key);
+                $qb->andWhere("$alias.$key = :$placeholder");
+                $queryBuilder->setParameter($placeholder, $value);
+            }
+
+            if (null === $previousQueryBuilder) {
+                $previousQueryBuilder = $qb;
+            } else {
+                $previousQueryBuilder->andWhere($qb->expr()->in($previousAlias, $qb->getDQL()));
+            }
+
+            $previousAlias = $alias;
+        }
+
+        $queryBuilder->where(
+            $queryBuilder->expr()->in('o', $previousQueryBuilder->getDQL())
+        );
```

Comment:
IMO this is the best way to do what we want because we force the SQL execution plan to go through indexes on doctrine identifiers. The other solution would be to use joins.

This transforms to (pseudo-dql):

```
SELECT thirdLevel
WHERE thirdLevel IN (
  SELECT thirdLevel FROM relatedDummies WHERE relatedDummies = ? AND relatedDummies IN (
    SELECT relatedDummies FROM Dummy WHERE Dummy = ?
  )
 )
```